src/main/java/org/mondemand/Client.java

```diff
@@ -1062,22 +1073,20 @@ public boolean traceMessage (String owner, String traceId, String message,
                                Map<String, String> context) {
     boolean ret = false;
     try {
-      /* FIXME: I'm sure there's a better way to do this but I am not the best
-       * java programmer.
-       */
-      ConcurrentHashMap<String,Context> tmp =
-        new ConcurrentHashMap<String,Context>();
-      for (Entry<String,String> entry : context.entrySet())
-        {
-          tmp.put(entry.getKey(),
-                  new Context(entry.getKey(), entry.getValue()));
+      List<Context> contextsList = new ArrayList<Context>(context.size() + 3);
+      contextsList.add(new Context(OWNER_KEY, owner));
+      contextsList.add(new Context(TRACE_KEY, traceId));
+      contextsList.add(new Context(MESSAGE_KEY, message));
+      for (Entry<String, String> entry : context.entrySet()) {
+        if (OWNER_KEY.equals(entry.getKey()) ||
+            TRACE_KEY.equals(entry.getKey()) ||
+            MESSAGE_KEY.equals(entry.getKey())) {
```

Comment:
I saw the FIXME and decided to take a look. This doesn't do the same thing as the original code but I'm pretty sure actual intent was to check for the OWNER_KEY, TRACE_KEY and MESSAGE_KEY as potential keys and not owner, traceId and message.